Sources/NIOSSL/SSLCertificate.swift

```diff
@@ -407,73 +397,108 @@ extension NIOSSLCertificate: Hashable {
-        self.nextIdx = 0
+        fileprivate init(_ rawCode: Int) {
+            self.rawValue = rawCode
+        }
+
+        fileprivate init(_ rawCode: Int32) {
+            self.init(Int(rawCode))
+        }
+        
+        public static let email = Self(GEN_EMAIL)
+        public static let dnsName = Self(GEN_DNS)
+        public static let ipAddress = Self(GEN_IPADD)
+    }
+
+    public struct Contents {
+        // only part of this type to keep a strong reference to the underlying storage of `buffer`
+        private let collection: _SubjectAlternativeNames
+        // lifetime automatically managed by `collection`
+        fileprivate let buffer: UnsafeBufferPointer<UInt8>
+        
+        internal init(collection: _SubjectAlternativeNames, buffer: UnsafeBufferPointer<UInt8>) {
+            self.collection = collection
+            self.buffer = buffer
+        }
     }
+    
+    internal enum IPAddress {
+        case ipv4(in_addr)
+        case ipv6(in6_addr)
+    }
+}
 
-    private func addressFromBytes(bytes: UnsafeBufferPointer<UInt8>) -> NIOSSLCertificate.IPAddress? {
-        switch bytes.count {
+extension _SubjectAlternativeName.Contents: RandomAccessCollection {
+    public var startIndex: Int { buffer.startIndex }
+    public var endIndex: Int { buffer.endIndex }
+    
+    public subscript(position: Int) -> UInt8 {
+        get {
+            buffer[position]
+        }
+    }
+}
+
+extension _SubjectAlternativeName.IPAddress {
+    init?(_ subjectAlternativeName: _SubjectAlternativeName) {
+        guard subjectAlternativeName.nameType == .ipAddress else {
+            return nil
+        }
+        switch subjectAlternativeName.contents.count {
         case 4:
-            let addr = bytes.baseAddress?.withMemoryRebound(to: in_addr.self, capacity: 1) {
+            let addr = subjectAlternativeName.contents.buffer.baseAddress?.withMemoryRebound(to: in_addr.self, capacity: 1) {
                 return $0.pointee
             }
             guard let innerAddr = addr else {
                 return nil
             }
-            return .ipv4(innerAddr)
+            self = .ipv4(innerAddr)
         case 16:
-            let addr = bytes.baseAddress?.withMemoryRebound(to: in6_addr.self, capacity: 1) {
+            let addr = subjectAlternativeName.contents.buffer.baseAddress?.withMemoryRebound(to: in6_addr.self, capacity: 1) {
                 return $0.pointee
             }
             guard let innerAddr = addr else {
                 return nil
             }
-            return .ipv6(innerAddr)
+            self = .ipv6(innerAddr)
         default:
             return nil
         }
     }
+}
 
-    func next() -> NIOSSLCertificate.AlternativeName? {
-        guard self.nextIdx < self.stackSize else {
-            return nil
-        }
+/// Collection of all Subject Alternative Names from a `NIOSSLCertificate`
+public final class _SubjectAlternativeNames: RandomAccessCollection {
```

Comment:
Can we wrap this `class` in a `struct`, where the `struct` is public and forwards to equivalent methods on the `class`?